zerver/views/message_send.py

```diff
@@ -346,3 +356,96 @@ def render_message_backend(
+
+
+@has_request_variables
+def report_message_backend(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    message_id: int = REQ(converter=to_non_negative_int, path_only=True),
+    reason: str = REQ(
+        default="other", str_validator=check_string_in(list(REPORT_MESSAGE_REASONS.keys()))
+    ),
+    explanation: str = REQ(default=""),
+) -> HttpResponse:
+    request_notes = RequestNotes.get_notes(request)
+    assert request_notes.log_data is not None
+    orig_message, ignored = access_message(user_profile, message_id, lock_message=False)
+    reason_str = REPORT_MESSAGE_REASONS[reason]
+    explanation = explanation.strip()
+    if reason == "other" and explanation == "":
+        raise JsonableError(_("For reason=other, an explanation must be provided."))
+
+    user_name, user_id = user_profile.full_name, user_profile.id
+    sender_name, sender_id = orig_message.sender.full_name, orig_message.sender.id
+    quoted_message = get_quoted_text(orig_message.content)
+    user = f"@_**{user_name}|{user_id}**"
+    sender = f"@_**{sender_name}|{sender_id}**"
+
+    # compute recipient
+    if orig_message.recipient.type == Recipient.STREAM:
+        stream = get_stream_by_id_in_realm(orig_message.recipient.type_id, user_profile.realm)
+        recipient = f"#**{stream.name}>{orig_message.topic_name()}**\n"
+    elif orig_message.recipient.type == Recipient.PERSONAL:
+        receiver = get_user_profile_by_id_in_realm(
+            orig_message.recipient.type_id, user_profile.realm
+        )
+        recipient = f"@**{receiver.full_name}|{receiver.id}**\n"
+    else:
+        recipient = " ".join([f"@**|{id}**" for id in get_huddle_user_ids(orig_message.recipient)])
+
+    # note: put original message at the end, so if the report_message gets truncated, it's from the original.
+    if explanation == "":
+        report_message = _(
+            "{user} reported a message sent by {sender} to {recipient} as containing **`{reason_str}`**.\n\nOriginal message:\n{quoted_message}"
+        ).format(
+            user=user,
+            sender=sender,
+            recipient=recipient,
+            reason_str=reason_str,
+            quoted_message=quoted_message,
+        )
+    else:
+        explanation = get_quoted_text(explanation)
+        report_message = _(
+            "{user} reported a message sent by {sender} to {recipient} as containing **`{reason_str}`** with these notes:\n{explanation}\nOriginal message:\n{quoted_message}"
+        ).format(
+            user=user,
+            sender=sender,
+            recipient=recipient,
+            reason_str=reason_str,
+            explanation=explanation,
+            quoted_message=quoted_message,
+        )
+
+    with override_language(user_profile.realm.default_language):
+        sender = get_system_bot(settings.NOTIFICATION_BOT, user_profile.realm_id)
+        topic = _("sent by {sender_name}").format(sender_name=sender_name)
```

Comment:
I think we want to thread abuse reports by the user being reported, not the user doing the report, so that multiple reports of the same target use end up together.